src/ray/gcs/gcs_server/gcs_node_manager.h

```diff
@@ -0,0 +1,70 @@
+#ifndef RAY_GCS_NODE_MANAGER_H
+#define RAY_GCS_NODE_MANAGER_H
+
+#include <ray/common/id.h>
+#include <ray/protobuf/gcs.pb.h>
+#include <ray/rpc/client_call.h>
+
+namespace ray {
+
+namespace gcs {
+class RedisGcsClient;
+/// The GcsNodeManager is responsible for detecting the death of raylet, it has all
+/// functionality of raylet_monitor.
+class GcsNodeManager {
+ public:
+  /// Create a GcsNodeManager.
+  ///
+  /// \param io_service The event loop to run the monitor on.
+  /// \param gcs_client The client of gcs to access/pub/sub data.
+  explicit GcsNodeManager(boost::asio::io_service &io_service,
+                          std::shared_ptr<gcs::RedisGcsClient> gcs_client);
+
+  /// Listen for heartbeats from Raylets and mark Raylets
+  /// that do not send a heartbeat within a given period as dead.
+  void Start();
+
+ protected:
+  /// Handle a heartbeat from a Raylet.
+  ///
+  /// \param client_id The client ID of the Raylet that sent the heartbeat.
+  /// \param heartbeat_data The heartbeat sent by the client.
+  void HandleHeartbeat(const ClientID &client_id,
+                       const rpc::HeartbeatTableData &heartbeat_data);
+
+  /// A periodic timer that fires on every heartbeat period. Raylets that have
+  /// not sent a heartbeat within the last num_heartbeats_timeout ticks will be
+  /// marked as dead in the client table.
+  void Tick();
+
+  /// Check that if any raylet is inactive due to no heartbeat for a period of time.
+  /// If found any, mark it as dead.
+  void DetectDeadClients();
```

Comment:
```suggestion
  void DetectDeadNodes();
```
The term `client` is ambiguous. I think we can use either `node` or `raylet`.